File: analytics/metrics.py
import numpy as np
import pandas as pd


def sharpe_ratio(returns, risk_free_rate=0.02):
    excess_returns = returns - risk_free_rate / 252
    if excess_returns.std() == 0:
        return 0
    return np.sqrt(252) * excess_returns.mean() / excess_returns.std()


def max_drawdown(equity_curve):
    peak = equity_curve['equity'].expanding().max()
    drawdown = (equity_curve['equity'] - peak) / peak
    return drawdown.min() * 100


def win_rate(trades):
    if not trades:
        return 0

    profits = []
    for i in range(0, len(trades) - 1, 2):
        if i + 1 < len(trades):
            buy = trades[i]
            sell = trades[i + 1]
            profit = sell['price'] - buy['price']
            profits.append(profit > 0)

    if not profits:
        return 0
    return sum(profits) / len(profits) * 100


def calculate_metrics(results):
    equity_df = results['equity_curve']
    trades = results['trades']

    if len(equity_df) < 2:
        return {}

    equity_df['returns'] = equity_df['equity'].pct_change()

    return {
        'total_return': results['return_pct'],
        'sharpe': sharpe_ratio(equity_df['returns'].dropna()),
        'max_drawdown': max_drawdown(equity_df),
        'win_rate': win_rate(trades),
        'num_trades': len(trades),
        'final_equity': results['final_equity']
    }


# TODO: add sortino ratio
# TODO: add calmar ratio

File: lib/analytics/metrics.ts
/**
 * Performance Metrics
 * Ported from analytics/metrics.py
 */

import { mean, std } from '../math/statistics';
import { expandingMax, pctChange, dropNaN } from '../math/rolling';
import { Metrics, Trade, EquityPoint, SimulationResult } from '../backtest/types';

/**
 * Calculate Sharpe Ratio (annualized)
 * Sharpe = sqrt(252) * mean(excess_returns) / std(excess_returns)
 */
export function sharpeRatio(returns: number[], riskFreeRate: number = 0.02): number {
  const cleanReturns = dropNaN(returns);
  if (cleanReturns.length === 0) return 0;

  const dailyRf = riskFreeRate / 252;
  const excessReturns = cleanReturns.map((r) => r - dailyRf);

  const stdDev = std(excessReturns);
  if (stdDev === 0) return 0;

  return Math.sqrt(252) * mean(excessReturns) / stdDev;
}

/**
 * Calculate Maximum Drawdown (percentage)
 * Returns the largest peak-to-trough decline
 */
export function maxDrawdown(equityCurve: EquityPoint[]): number {
  if (equityCurve.length === 0) return 0;

  const equities = equityCurve.map((e) => e.equity);
  const peak = expandingMax(equities);

  let maxDD = 0;
  for (let i = 0; i < equities.length; i++) {
    const dd = (equities[i] - peak[i]) / peak[i];
    if (dd < maxDD) maxDD = dd;
  }

  return maxDD * 100; // Return as percentage (negative)
}

/**
 * Calculate Win Rate (percentage of profitable trades)
 */
export function winRate(trades: Trade[]): number {
  if (trades.length === 0) return 0;

  const profits: boolean[] = [];

  // Pair up buy/sell trades
  for (let i = 0; i < trades.length - 1; i += 2) {
    if (i + 1 < trades.length) {
      const buy = trades[i];
      const sell = trades[i + 1];
      if (buy.type === 'buy' && sell.type === 'sell') {
        profits.push(sell.price > buy.price);
      }
    }
  }

  if (profits.length === 0) return 0;
  return (profits.filter((p) => p).length / profits.length) * 100;
}

/**
 * Calculate returns from equity curve
 */
export function calculateReturns(equityCurve: EquityPoint[]): number[] {
  const equities = equityCurve.map((e) => e.equity);
  return dropNaN(pctChange(equities, 1));
}

/**
 * Calculate all metrics from simulation result
 */
export function calculateMetrics(result: SimulationResult): Metrics {
  const { equityCurve, trades, finalEquity, returnPct, numTrades } = result;

  if (equityCurve.length < 2) {
    return {
      totalReturn: 0,
      sharpe: 0,
      maxDrawdown: 0,
      winRate: 0,
      numTrades: 0,
      finalEquity: 10000,
    };
  }

  const returns = calculateReturns(equityCurve);

  return {
    totalReturn: returnPct,
    sharpe: sharpeRatio(returns),
    maxDrawdown: maxDrawdown(equityCurve),
    winRate: winRate(trades),
    numTrades,
    finalEquity,
  };
}

/**
 * Calculate buy-and-hold returns for comparison
 */
export function buyAndHoldReturn(
  prices: number[],
  initialCapital: number = 10000
): number {
  if (prices.length < 2) return 0;
  const startPrice = prices[0];
  const endPrice = prices[prices.length - 1];
  return ((endPrice - startPrice) / startPrice) * 100;
}

/**
 * Generate buy-and-hold equity curve
 */
export function buyAndHoldEquity(
  equityCurve: EquityPoint[],
  initialCapital: number = 10000
): EquityPoint[] {
  if (equityCurve.length === 0) return [];

  const startPrice = equityCurve[0].price;
  const shares = initialCapital / startPrice;

  return equityCurve.map((point) => ({
    date: point.date,
    equity: shares * point.price,
    price: point.price,
  }));
}
